### ISSUE TITLE
COM ribbon problems and my experimental version

### ISSUE THREAD

**Issue by fandrei** [Excel-DNA/ExcelDna#107]
Hello Govert,
Here is an experimental version I created: https://github.com/fandrei/ExcelDnaExperimenting
It does 3 things:
- added custom logging support, so ExcelDna logs can be captured by user's addin and sent to a log tracker sever (for instance)
- delay unregistration of COM addins
- use volatile registry keys for COM addin registration, so they won't stay forever even if Excel crashes

The addin with this version lets the COM addin/Ribbon work on some PC where it didn't; however, these PCs still are reporting this COM error: 

```
ExcelDna: ComAddIn	The Ribbon/COM add-in helper required by add-in Shooju could not be registered.
This is an unexpected error.
```

Could you take a look at this code?

**Comment by govert** [comment]
@fandrei Thank you - I'll have a look.

Would it not be possible to do your logging with the normal logging infrastructure (by just writing your TraceListener)? See the various configuration examples here: https://github.com/Excel-DNA/ExcelDna/wiki/Diagnostic-Logging

I've never heard of volatile registry keys, but they seem like a good idea here - I'll have to read a bit more about that.

**Comment by fandrei** [comment]
@govert, 
> Would it not be possible to do your logging with the normal logging infrastructure (by just writing your TraceListener)?

This would intercept all messages written to the system diag log, including our own; and this would created a recursive loop of messages and required some complex logic to avoid it.

**Comment by govert** [comment]
@fandrei I don't understand your concern - you could configure your TraceListener to listen to whatever TraceSources you want to, and can filter messages as you need.

I'm just not keen to support two parallel logging approaches.

**Comment by fandrei** [comment]
@govert, possibly, but this is not of top priority. The COM addin loading problem is most important here.

**Comment by govert** [comment]
@fandrei Do you know if you just use the volatile registry keys, but leave the key removal as it was (directly after the `AddIn.Connect(true)` call), does it still work?

There are two things that don't make sense to me:
* How do you get the exception and show the message, if the add-in load was successful? Maybe you can add the `Exception.Message` to the displayed error message - perhaps it has something to do with the registry deletes themselves...?
* How does deleting or not deleting the registry keys after the add-in is loaded (after `AddIn.Connect(true) has been called) affect whether the add-in loads or not?

**Comment by fandrei** [comment]
@govert, 

> Do you know if you just use the volatile registry keys, but leave the key removal as it was (directly after the AddIn.Connect(true) call), does it still work?

It's difficult to test. Only some users have this problem and I can't reproduce it. It seems that we have to use guesswork here.

> How do you get the exception and show the message, if the add-in load was successful? 

That error message is reported by the logger, but the addin load seems to continue despite this exception.

> How does deleting or not deleting the registry keys after the add-in is loaded (after `AddIn.Connect(true) has been called) affect whether the add-in loads or not?

With the original code, Excel-DNA shows the "unspecified error" when starting and Ribbon doesn't show up. With these modifications, the same error messages is reported, but Ribbon starts and works correctly.

**Comment by govert** [comment]
@fandrei My understanding is that the `AddIn.Connect(true)` call is where the add-in gets loaded. The changes to when the registry entries get removed only affects code what happens after this. So I can't see (by reasoning from the code and my understanding) how that change can make a difference to the add-in loading. However, the volatile change does make a difference to how the registry entries are written, before the add-in is loaded. So that might have an effect, given my understanding of the timing of the loading.

Anyway, I think we'll need to work on a different debugging approach for this problem.

**Comment by fandrei** [comment]
@govert, I wonder if it can just start loading and return before the completion.

Volatile keys shouldn't change anything from the perspective of the consumer application. They just get deleted automatically by the system on the next reboot.

**Comment by govert** [comment]
> I wonder if it can just start loading and return before the completion.
Then it couldn't throw throw the exception which causes that message to be shown.

If the volatile keys don't make a difference, then we're back to the situation where the code and the behaviour don't match.

Could you perhaps have multiple COM add-in objects being loaded - e.g. if you have a ribbon and loaded some CTP? That would explain how the ribbon could load and still show the error.

Otherwise you have old configuration settings stuck on those machines, confusing the behaviour we see.

**Comment by fandrei** [comment]
@govert, CTP?

**Comment by govert** [comment]
CTP = Custom Task Pane. The Excel-DNA support for these load a second helper COM add-in.
The reason I ask is that you have a situation where the COM add-in loads, and you still get an error. I don't know what the source of the error is in this case.

Maybe you can capture a stack trace which might tell us more about this case - just add the whole `Exception.ToString()` to the message that is displayed.

I have to note that although other users have reported issues with the Ribbon add-in loading, these have generally been caused by the COM helper add-in being marked Disabled by Excel due to unhandled exceptions from methods in the the `ExcelRibbon`-derived class. The few other registry-related issues seem to have been addressed by the last version. So the problems you see might be specific to something in your environment or add-in.

**Comment by fandrei** [comment]
@govert, no we don't have any. There is only one Ribbon tab in this addin.

The exception was the same:
```
ExcelDna.Integration Information: 4 : LoadComAddIn exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Runtime.InteropServices.COMException: Unspecified error (Exception from HRESULT: 0x80004005 (E_FAIL))
   --- End of inner exception stack trace ---
   at System.RuntimeType.InvokeDispMethod(String name, BindingFlags invokeAttr, Object target, Object[] args, Boolean[] byrefModifiers, Int32 culture, String[] namedParameters)
   at System.RuntimeType.InvokeMember(String name, BindingFlags bindingFlags, Binder binder, Object target, Object[] providedArgs, ParameterModifier[] modifiers, CultureInfo culture, String[] namedParams)
   at ExcelDna.Integration.ExcelComAddInHelper.LoadComAddIn(ExcelComAddIn addIn) with /K in CommandLine
ExcelDna.Integration Verbose: 1 : Exit Integration.DnaLibraryAutoOpen
```

I wonder what registry keys could cause problems with Ribbon loading?

**Comment by govert** [comment]
@fandrei And you see this same exception even when the Ribbon does load successfully? That's the mystery.

**Comment by fandrei** [comment]
@govert, Hello. We've got a couple more exceptions, these are different.
```
###ExcelDna: ComAddIn	ExcelRibbon.GetCustomUI - Invalid RibbonID for Excel. RibbonID: Microsoft.Mso.IMLayerUI
```
and
```
###ExcelDna: ComAddIn	LoadComAddIn exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Runtime.InteropServices.COMException: Class not registered (Exception from HRESULT: 0x80040154 (REGDB_E_CLASSNOTREG))
   --- End of inner exception stack trace ---
   at System.RuntimeType.InvokeDispMethod(String name, BindingFlags invokeAttr, Object target, Object[] args, Boolean[] byrefModifiers, Int32 culture, String[] namedParameters)
   at System.RuntimeType.InvokeMember(String name, BindingFlags bindingFlags, Binder binder, Object target, Object[] providedArgs, ParameterModifier[] modifiers, CultureInfo culture, String[] namedParams)
   at System.Type.InvokeMember(String name, BindingFlags invokeAttr, Binder binder, Object target, Object[] args, CultureInfo culture)
   at ExcelDna.Integration.ExcelComAddInHelper.LoadComAddIn(ExcelComAddIn addIn)
```
Do they say anything to you?

**Comment by govert** [comment]
@fandrei Is this running with your "experimental version"?

The first error - where the host is passing in "Microsoft.Mso.IMLayerUI" as the ribbon ID is related to Outlook trying to load the add-in. Maybe this happens when an Excel workbook in embedded in a mail message somehow? There are some problems when Excel is running in an embedded mode in another application - it's not a case I understand well, but it's unlikely we can make the add-in work well in this case due to some weird Excel limitations. As best we can keep the add-in (or parts of it) disabled without crashing.

The second is harder to pinpoint. One would need to check the usual issues again:
* Are there any Disabled items?
* Is Excel running in protected mode?
* Is it running in some virtualized (Citrix) or otherwise locked down environment where some parts of the user registry cannot be written to?

**Comment by fandrei** [comment]
@govert, yes

> Are there any Disabled items?
> Is Excel running in protected mode?

I wonder if it's possible to detect this programmatically from the addin?

**Comment by govert** [comment]
@fandrei yes?

**Comment by fandrei** [comment]
@govert, how can I get the name of the RTD addin used by the current addin?
Do you know how to detect programmatically if Excel is in the protected mode?

**Comment by govert** [comment]
@fandrei Were there Disabled add-in?

**Comment by fandrei** [comment]
@govert, no, AFAIK. However, I don't have access to that PC. I can't say exactly unless I add code to check this programmatically and log it.

**Comment by govert** [comment]
@fandrei Was this running your experimental version?

**Comment by fandrei** [comment]
@govert, yes

**Comment by govert** [comment]
It's not actually Excel that runs in protected mode, but a particular window that is a protected view - the COM type is a ProtectedViewWindow or something. But when this is the only window that is open, I think Excel behaves a bit weird. There is a fix lined up in v0.34 for one cases where this caused a problem (it caused ExcelDnaUtil.Application to fail). But I remember an issue in the past with this related to the ribbon - not sure it that case would be fixed too.

The disabled items are written in the registry somewhere. I don't have the details.

**Comment by fandrei** [comment]
@govert, thanks.

> The disabled items are written in the registry somewhere. I don't have the details.

The main problem is getting the name of the RTD COM addin. AFAIK, ExcelDna generates this name somehow?

**Comment by govert** [comment]
You have a persistent RTD server, so the name is the ProgId you assigned.

**Comment by govert** [comment]
Can I close this issue - I think we've drifted off a bit?
The Google group is better for general discussion.

**Comment by fandrei** [comment]
@govert, ok

**Comment by govert** [comment]
@fandrei You're welcome to start a discussion on the Google group about the COM registration details.

**Comment by govert** [comment]
This issue was caused by the Office "AutomationSecurity" setting, and is addressed by commit 684c13af49dfe9d77b2e1cff1c6fdd357b5e6323
